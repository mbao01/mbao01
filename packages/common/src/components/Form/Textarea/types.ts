import type { ReactNode } from "react";
import type { VariantProps } from "../../../libs";
import { getTextareaClasses } from "./constants";

export type TextareaProps = Omit<
  React.TextareaHTMLAttributes<HTMLTextAreaElement>,
  "size" | "ref"
> &
  VariantProps<typeof getTextareaClasses> & {
    label?: ReactNode;
  };

export type TextareaLabelProps = {
  children: ReactNode;
};
