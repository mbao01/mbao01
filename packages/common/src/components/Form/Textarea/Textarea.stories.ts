import type { Meta, StoryObj } from "@storybook/react";

import { Textarea } from "./Textarea";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Textarea,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
    // backgrounds: [{ name: 'dark background', value: '#000', default: true }],
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Textarea>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    placeholder: "First name",
  },
};

export const BorderedTextarea: Story = {
  args: {
    outline: true,
    placeholder: "Full name",
  },
};

export const PrimaryTextarea: Story = {
  args: {
    variant: "primary",
    placeholder: "Full name",
  },
};

export const TinyTextarea: Story = {
  args: {
    size: "xs",
    variant: "accent",
    placeholder: "Full name",
  },
};

export const DisabledTextarea: Story = {
  args: {
    disabled: true,
  },
};
