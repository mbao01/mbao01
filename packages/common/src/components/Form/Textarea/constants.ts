import { cva } from "../../../libs";

export const getTextareaClasses = cva("textarea rounded-md transition-all duration-100", {
  variants: {
    variant: {
      default: "bg-transparent",
      accent: "textarea-accent",
      error: "textarea-error",
      ghost: "textarea-ghost",
      info: "textarea-info",
      primary: "textarea-primary",
      secondary: "textarea-secondary",
      success: "textarea-success",
      warning: "textarea-warning",
    },
    outline: {
      false: "textarea-ghost",
    },
    wide: {
      true: "w-full",
    },
    size: {
      xs: "textarea-xs",
      sm: "textarea-sm",
      md: "textarea-md",
      lg: "textarea-lg",
    },
  },
  compoundVariants: [
    {
      size: undefined,
      className: "min-h-fit h-10",
    },
    {
      outline: undefined,
      className: "textarea-ghost",
    },
    {
      variant: undefined,
      outline: true,
      className: "border-neutral-content",
    },
    {
      variant: "default",
      outline: true,
      className: "border-base-content",
    },
  ],
});
