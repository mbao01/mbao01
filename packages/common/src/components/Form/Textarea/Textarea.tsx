import * as React from "react";
import { type TextareaProps } from "./types";
import { cn } from "../../../utilities";
import { getTextareaClasses } from "./constants";

const Textarea = React.forwardRef<HTMLTextAreaElement, TextareaProps>(
  ({ size, variant, wide, outline, className, ...props }, ref) => {
    return (
      <textarea
        ref={ref}
        className={cn(
          getTextareaClasses({ size, wide, variant, outline }),
          className
        )}
        {...props}
      />
    );
  }
);
Textarea.displayName = "Textarea";

export { Textarea };
