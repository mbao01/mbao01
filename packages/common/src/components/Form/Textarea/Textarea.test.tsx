import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it } from "vitest";
import { Textarea } from "./Textarea";

describe("Textarea", () => {
  it("renders with the correct initial state", () => {
    const { asFragment } = render(<Textarea name="username" defaultValue="John Doe" />);

    expect(screen.getByDisplayValue("John Doe")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("handles user textarea correctly", async () => {
    const user = userEvent.setup();

    const { asFragment } = render(<Textarea name="username" defaultValue="John Doe" />);
    const textarea = screen.getByDisplayValue("John Doe");

    await user.clear(textarea);
    await user.type(textarea, "Ayomide B.");

    expect(textarea).toHaveValue("Ayomide B.");
    expect(asFragment()).toMatchSnapshot();
  });

  it("prevent user textarea when disabled", async () => {
    const user = userEvent.setup();

    const { asFragment } = render(<Textarea name="username" defaultValue="John Doe" disabled />);
    const textarea = screen.getByDisplayValue("John Doe");

    await user.type(textarea, "Ayomide B.");

    expect(textarea).toHaveValue("John Doe");
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    "primary",
    "secondary",
    "accent",
    "ghost",
    "info",
    "success",
    "warning",
    "error",
  ] as const)("has %s variant", (variant) => {
    const value = `${variant} button`;
    const { asFragment } = render(<Textarea name="name" defaultValue={value} variant={variant} />);

    expect(screen.getByDisplayValue(value)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    { size: "xs", name: "tiny" },
    { size: "sm", name: "small" },
    { size: "md", name: "base" },
    { size: "lg", name: "large" },
  ] as const)("has $name ($size) textarea", ({ size, name }) => {
    const value = `${name} textarea`;
    const { asFragment } = render(<Textarea name="name" defaultValue={value} size={size} />);

    expect(screen.getByDisplayValue(value)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("has a floating label", () => {
    const { asFragment } = render(<Textarea label="Your work experience" />);

    expect(screen.getByLabelText("Your work experience")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
});
