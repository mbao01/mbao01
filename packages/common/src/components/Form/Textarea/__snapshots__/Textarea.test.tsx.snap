// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Textarea > handles user textarea correctly 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 min-h-fit h-10 textarea-ghost"
    name="username"
  >
    John Doe
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has 'base' ('md') textarea 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 textarea-md textarea-ghost"
    name="name"
  >
    base textarea
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has 'large' ('lg') textarea 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 textarea-lg textarea-ghost"
    name="name"
  >
    large textarea
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has 'small' ('sm') textarea 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 textarea-sm textarea-ghost"
    name="name"
  >
    small textarea
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has 'tiny' ('xs') textarea 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 textarea-xs textarea-ghost"
    name="name"
  >
    tiny textarea
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has a label 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 min-h-fit h-10 textarea-ghost"
    name="username"
  >
    Username
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has accent variant 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 textarea-accent min-h-fit h-10"
    name="name"
  >
    accent button
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has error variant 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 textarea-error min-h-fit h-10"
    name="name"
  >
    error button
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has ghost variant 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 textarea-ghost min-h-fit h-10"
    name="name"
  >
    ghost button
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has info variant 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 textarea-info min-h-fit h-10"
    name="name"
  >
    info button
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has primary variant 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 textarea-primary min-h-fit h-10"
    name="name"
  >
    primary button
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has secondary variant 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 textarea-secondary min-h-fit h-10"
    name="name"
  >
    secondary button
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has success variant 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 textarea-success min-h-fit h-10"
    name="name"
  >
    success button
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > has warning variant 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 textarea-warning min-h-fit h-10"
    name="name"
  >
    warning button
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > prevent user textarea when disabled 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 min-h-fit h-10 textarea-ghost"
    disabled=""
    name="username"
  >
    John Doe
  </textarea>
</DocumentFragment>
`;

exports[`Textarea > renders with the correct initial state 1`] = `
<DocumentFragment>
  <textarea
    class="textarea rounded-md transition-all duration-100 min-h-fit h-10 textarea-ghost"
    name="username"
  >
    John Doe
  </textarea>
</DocumentFragment>
`;
